Public developer guide (#821)
Infleqtion/client-superstaq: docs/source/resources/developer_guide.rst

Reviewer: `check/` should be replaced with `checks/` everywhere it appears in this file

Author: updated!

|       gaussian_pulse = library.gaussian(...)     |       gaussian_pulse = library.gaussian(...)     |
|       pulse.play(gaussian_pulse)                 |       pulse.play(gaussian_pulse)                 |
+--------------------------------------------------+--------------------------------------------------+

Exceptions can be made if the explicit name is extremely long (e.g. ``from qiskit.providers import JobStatus as qjs``) or if the name is used extremely frequently and has low risk of name clash (e.g. ``from typing import Iterable, Union``).

Productivity Tips
=================
* For command-line python testing, use the ``ipython`` shell instead of normal python. ``ipython`` is like Jupyter but within your terminal.
* For either ``ipython`` or Jupyter, consider starting with

  >>> %load_ext autoreload
  >>> %autoreload 2

  This will enable autoreload.
* Also recommended: 

.. code-block:: bash

   alias py='ipython'
   alias pyr="ipython --InteractiveShellApp.extensions 'autoreload' --InteractiveShellApp.exec_lines '%autoreload 2'"
alias pygrep="grep -r --color=auto --include='*.py'"

> (That last one, ``pyr``, runs an ipython shell with autoreloading already set up.)

Reviewer: nit: `pyr` isn't the "last one" anymore (maybe just reorder them?)

===============
Developer Guide
===============

This is the developer guide for Superstaq. If you have questions, please email superstaq@infleqtion.com or reach out on our `Slack workspace <https://join.slack.com/t/superstaq/shared_invite/zt-1wr6eok5j-fMwB7dPEWGG~5S474xGhxw>`_.

.. contents:: **Table of Contents**
   :depth: 3
   :local:
   :backlinks: none

Developer Workflows
===================

Installation
------------
Please visit our `README <https://github.com/Infleqtion/client-superstaq/blob/main/README.md#installation-for-development>`_ for developer installation instructions.

Modifying Superstaq
-------------------

Push Edits and Make a Pull Request for Review
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> #. Update the main branch (e.g., ``git pull``) and update relevant dependencies (``python3 -m pip install -e .``, ``python3 -m pip install <general-superstaq, qiskit-superstaq, cirq-superstaq. --upgrade``).

Reviewer: this install line isn't quite right - should be something like,
```suggestion
#. Update the main branch (e.g., ``git pull``)
#. Install/update relevant dependencies. To work on a single package, navigate to relevant subdirectory (e.g. `cd cirq-superstaq`) and run ``python3 -m pip install -e .``. Alternatively, install all ``client-superstaq`` packages by navigating to the repository root and running ``python3 -m pip install -e ./checks-superstaq -e ./general-superstaq -e ./qiskit-superstaq -e ./cirq-superstaq -e ./supermarq-benchmarks`` (in that order).
```
(though probably cleaned up a bit)

Author: Add the ^ language with some formatting - let me know how it looks and if you want me to update anything!

Reviewer: perfect, thanks!

    check/build_docs.py  # Ensures docs can be built. This will fail for e.g. incorrectly formatted code blocks in docstrings.
    check/all_.py  # Runs all the non-integration checks described above.

By default, all test scripts will consider any tracked file in the current repository (meaning that new files will not be checked until they've been added to the repo via ``git add``). Passing ``-i`` or ``--incremental`` to any check will limit its scope to just locally modified files. Test scripts can also be passed individual files or subdirectories or prevented from checking specific files using ``-x <path>`` or ``--exclude <path>``.

Style Guide
===========

Nomenclature and Abbreviations
------------------------------

Operations can be referred to as ``op``.

``target`` should always be used to refer to device names (type ``str``) in the Superstaq "<vendor>_<device>_<type>" format, e.g. "ibmq_qasm_simulator".  The same will be used accross all clients. ``backend_name`` should be used for vendor-specific names of particular hardware backends, and ``backend`` for actual instances thereof.

> Circuits

Reviewer: this is cirq-specific - maybe we should change the header to
```suggestion
Cirq
```
?

|                                                  |                                                  |
|.. code-block::                                   |.. code-block::                                   |
|                                                  |                                                  |
|   # BAD: confusing that "pulse" is a module      |   # GOOD                                         |
|   from qiskit import pulse                       |   import qiskit                                  |
|   with pulse.build(backend) as program:          |   with qiskit.pulse.build(backend) as program:   |
|       gaussian_pulse = library.gaussian(...)     |       gaussian_pulse = library.gaussian(...)     |
|       pulse.play(gaussian_pulse)                 |       pulse.play(gaussian_pulse)                 |
+--------------------------------------------------+--------------------------------------------------+

Exceptions can be made if the explicit name is extremely long (e.g. ``from qiskit.providers import JobStatus as qjs``) or if the name is used extremely frequently and has low risk of name clash (e.g. ``from typing import Iterable, Union``).

> 
> We allow ``web`` to import ``superstaq``, but not the reverse. Similarly, we allow ``dev_tools`` to import ``web``, but not the reverse.

Reviewer: ```suggestion
```

Productivity Tips
=================
* For command-line python testing, use the ``ipython`` shell instead of normal python. ``ipython`` is like Jupyter but within your terminal.
* For either ``ipython`` or Jupyter, consider starting with

  >>> %load_ext autoreload
  >>> %autoreload 2

  This will enable autoreload.

> * Superstaq/dev_tools/pygrep.sh contains a nice script for grepping python files. Recommendation: alias ``p`` to this script (by adding ``alias p='/path/to/Superstaq/dev_tools/pygrep.sh'`` to ~/.zprofile or wherever aliases should be stored). Then you can type something like ``p AQTDevice``

Reviewer: ```suggestion
```
(`pygrep.sh` isn't in client-superstaq)

* Also recommended: 

.. code-block:: bash

>    alias ss='cd /path/to/Superstaq'

Reviewer: ```suggestion
```

   alias py='ipython'
   alias pyr="ipython --InteractiveShellApp.extensions 'autoreload' --InteractiveShellApp.exec_lines '%autoreload 2'"

> 

Reviewer: ```suggestion
alias pygrep="grep -r --color=auto --include='*.py'"

```
(this does the same thing as `pygrep.sh`)

(That last one, ``pyr``, runs an ipython shell with autoreloading already set up.)

> 
> * View your AWS credentials with ``cat ~/.aws/credentials``, and store them in environment variables with, e.g., ``export AWS_ACCESS_KEY_ID=...``.

Reviewer: ```suggestion
```
